lib/Dialect/ESI/runtime/cpp/lib/backends/Xrt.cpp

```diff
@@ -100,6 +115,47 @@ class XrtMMIO : public MMIO {
 };
 } // namespace
 
+namespace {
+/// Host memory service specialized to XRT.
+class XrtHostMem : public HostMem {
+public:
+  XrtHostMem(::xrt::device &device, int32_t memoryGroup)
+      : device(device), memoryGroup(memoryGroup){};
+
+  struct XrtHostMemRegion : public HostMemRegion {
+    XrtHostMemRegion(::xrt::device &device, std::size_t size,
+                     HostMem::Options opts, int32_t memoryGroup) {
+      bo = ::xrt::bo(device, size, ::xrt::bo::flags::host_only, memoryGroup);
+      // Map the buffer into application memory space so that the application
+      // can use it just like any memory -- no need to use bo::write.
+      ptr = bo.map();
+    }
+    virtual void *getPtr() const override { return ptr; }
+    /// On XRT platforms, the pointer which the device sees is different from
+    /// the pointer the user application sees.
+    virtual void *getDevicePtr() const override { return (void *)bo.address(); }
+    virtual std::size_t getSize() const override { return bo.size(); }
+    /// It is recommended to use 'sync' to flush the caches before executing any
+    /// DMA.
+    virtual void flush() override { bo.sync(XCL_BO_SYNC_BO_TO_DEVICE); }
+
```

Comment:
Is there any sense in adding 'dirty'-flags to `HostMemRegion` s.t. when a `write` has been performed to it, `HostMem` implementations can check/verify that flush has been called?

I'm just a bit sceptical about the "it's recommended" phrase (although that probably originates from xrt documentation and not you).  I can accept "it's recommended" if it's purely performance related, but if not, it either is required or it isn't.

Reply:
There isn't a 'write' method on HostMemRegion. Standard use case is that the application writes directly to the host pointer. Avoids a memcpy, which as you well know is a killer on the host. So I have no way of (quickly) determining whether the buffer is dirty.

As for it's recommended... certain platforms will require it, and certain platforms won't. In the interest of applications writing platform-independent code, I wrote that it's recommended. As for Xrt, I found some documentation saying:

```
XRT Native API: xrt::bo::sync()) should be used. Though these API will not do any DMA operation, they are used for Cache Invalidate/Flush as the application works on the Cache memory.
```

so it sounds like it's required. I'll update the comment.

Reply:
Actually, I'm surprised that it's not required for the Catapult based platforms on x86. But I don't fully understand how this new user-pointer translation support works. It must go through the cache system.

Reply:
Said documentation: https://xilinx.github.io/XRT/master/html/hm.html